sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVOptions.scala

```diff
@@ -148,7 +148,28 @@ class CSVOptions(
   // A language tag in IETF BCP 47 format
   val locale: Locale = parameters.get("locale").map(Locale.forLanguageTag).getOrElse(Locale.US)
 
-  val dateFormatInRead: Option[String] = parameters.get("dateFormat")
+  /**
+   * Infer columns with all valid date entries as date type (otherwise inferred as timestamp type).
+   * Disabled by default for backwards compatibility and performance. When enabled, date entries in
+   * timestamp columns will be cast to timestamp upon parsing. Not compatible with
+   * legacyTimeParserPolicy == LEGACY since legacy date parser will accept extra trailing characters
+   */
+  val inferDate = {
+    val inferDateFlag = getBool("inferDate")
+    if (SQLConf.get.legacyTimeParserPolicy == LegacyBehaviorPolicy.LEGACY && inferDateFlag) {
```

Comment:
instead of failing here, shall we directly use `Iso8601DateFormatter` to infer date type?

Reply:
Does the `Iso8601DateFormatter` properly parse legacy date types? I'm inclined to believe it doesn't since the `getFormatter` returns a special `LegacyDateFormatter` when the `legacyTimeParserPolicy == LEGACY`

https://github.com/apache/spark/blob/9c5c21ccc9c7f41a204a738cd540c14793ffc8cb/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateFormatter.scala#L174-L188

cc @bersprockets @HyukjinKwon

Reply:
Yeah, I don't think legacy mode support this. Let's just throw an exception for now.

Reply:
My point is, we can use a dedicated datetime parser when inferring date types, which does not need to respect `legacyTimeParserPolicy`, so there is no need of failing for unsupported parser policy.

Reply:
In the [most recent commit](https://github.com/apache/spark/pull/36871/commits/e1170d0ee2027d810d2b23243602de147748838b), I implemented the suggestion from @cloud-fan to use the non-legacy parser for inference and allowing `inferDate=true` with `legacyTimeParserPolicy = LEGACY`

Reply:
>to always use the non-legacy parser for inference and allowing inferDate=true with legacyTimeParserPolicy = LEGACY. What do you think?

It's a little weird that you can specify `inferDate=true` with `legacyTimeParserPolicy=LEGACY`, yet Spark won't properly infer the type of legacy dates, e.g.:
```
scala> sql("set spark.sql.legacy.timeParserPolicy=legacy")
res0: org.apache.spark.sql.DataFrame = [key: string, value: string]

scala> // 1500-02-29 is a legacy date

scala> val csvInput = Seq("1425-03-22", "2022-01-01", "1500-02-29").toDS
csvInput: org.apache.spark.sql.Dataset[String] = [value: string]

scala> // infers as string

scala> spark.read.options(Map("inferSchema" -> "true", "inferDate" -> "true", "dateFormat" -> "yyyy-MM-dd")).csv(csvInput).printSchema
root
 |-- _c0: string (nullable = true)


scala> // if you specify a schema, it parses just fine, although 1500-02-29 becomes 1500-03-01 (expected)

scala> spark.read.schema("dt date").options(Map("dateFormat" -> "yyyy-MM-dd")).csv(csvInput).show(false)
+----------+
|dt        |
+----------+
|1425-03-22|
|2022-01-01|
|1500-03-01|
+----------+

scala> // remove the legacy date from the input...

scala> val csvInput = Seq("1425-03-22", "2022-01-01").toDS
csvInput: org.apache.spark.sql.Dataset[String] = [value: string]

scala> // .. and *then* Spark will infer date type

scala> spark.read.options(Map("inferSchema" -> "true", "inferDate" -> "true", "dateFormat" -> "yyyy-MM-dd")).csv(csvInput).printSchema
root
 |-- _c0: date (nullable = true)
```
If the user doesn't have any legacy dates in their input, it will work just fine, but then I am not sure why the user would be specifying `legacyTimeParserPolicy=LEGACY`.

Reply:
@HyukjinKwon Do you have a suggestion for whether we can merge with this latest change or if we should roll it back?

Reply:
I am fine with rolling back